### PR TITLE
Update: no-useless-escape / exact locations of errors (fixes #5751)

### LINKED ISSUE #5751: no-useless-escape could flag exact location of invalid escape sequence
This is an enhancement request for `no-useless-escape`. Right now, we report on the literal node, which results in an error message location that points to the beginning of the string or RegExp; it would be awesome if we could take into account the offset of the offending character so that we could give a more precise location.

The issue is more obvious when you have multiple problems in your literal:

```js
/* eslint no-useless-escape: "error" */
var myString = "\ab\c";

// error 1:16 - invalid escape \a (ideally this would say 1:17)
// error 1:16 - invalid escape \c (ideally this would say 1:20)
```

Please let me know if I can make this more clear.

### PR DESCRIPTION
**Pull request by onurtemizkan**
fixes: #5751

### CODE REVIEW

**Comment by onurtemizkan** [comment]
@eslintbot wake up :laughing:

File: lib/rules/no-useless-escape.js [hunk 1]
```diff
@@ -61,20 +61,38 @@ var VALID_REGEX_ESCAPES = [
 module.exports = function(context) {
 
     /**
+    * Returns location of the unnecessary escape
+    *
+    * @param {ASTNode} node - the node containing the unnecessary escape
+    * @param {int} indexInNode - the index of unnecessary escape inside node
+    * @returns {Object} - the location of the unnecessary escape
+    */
+    function getLocation(node, indexInNode) {
```
Comment (BYK) [inline lib/rules/no-useless-escape.js]: I don't think we need a function for this since it is pretty straightforward and used only in one place.

File: lib/rules/no-useless-escape.js [hunk 2]
```diff
@@ -95,10 +114,12 @@ module.exports = function(context) {
             return;
         }
 
-        (node.raw.match(/\\[^\d]/g) || []).forEach(validate, {
-            "escapes": nodeEscapes,
-            "node": node
-        });
+        while ((match = pattern.exec(node.raw))) {
```
Comment (BYK) [inline lib/rules/no-useless-escape.js]: Is this change necessary?

Comment (platinumazure) [inline lib/rules/no-useless-escape.js]: @BYK What context object is created in each iteration? Assuming the pattern has the "g" flag, this is a known pattern for evaluating the code one match at a time, and as far as I'm aware no waste is created.

Comment (BYK) [inline lib/rules/no-useless-escape.js]: @platinumazure this one:

```
{
    "node": node
    "escapes": nodeEscapes,
}
```

I actually think the following would be better and more efficient:

```
var validator = validate.bind({ escapes, nodeEscapes, node: node });
while ((match = pattern.exec(node.raw))) { validator(match); }
```

Comment (platinumazure) [inline lib/rules/no-useless-escape.js]: Oh, that context object. Yes, I agree that could be more efficient, still
fairly readable, and worth doing.

File: lib/rules/no-useless-escape.js (hunk 1 above)
Comment (mysticatea) [inline lib/rules/no-useless-escape.js]: But I like this function :smile:

File: lib/rules/no-useless-escape.js (hunk 2 above)
Comment (mysticatea) [inline lib/rules/no-useless-escape.js]: I think better using just a normal call.

    validate(nodeEscapes, node, match);

Comment (BYK) [inline lib/rules/no-useless-escape.js]: @mysticatea oh yeah, good call.

**Comment by onurtemizkan** [comment]
Updated :smile:

**Comment by mysticatea** [comment]
LGTM, thank you!

**Comment by nzakas** [comment]
Lgtm
